### PR TITLE
Fix token types to use LSP token types

### LINKED ISSUE #366: Rascal LSP uses textmate token scopes where it should use SemanticTokenTypes for the syntax highlighting.
**Describe the bug**

Rascal LSP is reporting the wrong kind of semantic token types in the legend. This makes it a hit and miss which ones are properly highlighted.

**To Reproduce**

Annotate a production with `@Category{constant.numeric}` and it doesn't get the colors of numeric contants of the theme, while `string` as category does work.

**Analysis**

I think the [documentation on this feature](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokenTypes) has been improved, it used to be less clear what LSP expected, I've read some old GH issues discussing the closed aspect of these TokenTypes. The set of valid token type is limited to the enum mentioned in the LSP spec, which is a subset of the [API in VS Code](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#standard-token-types-and-modifiers)

Now why did it even work? [We were passing sublime scopes](https://github.com/usethesource/rascal-language-servers/blob/9a7d76f033365d9077662faf279bd6ab1a12c463/rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java#L148-L349
)  like `entity.name` and `variable.function`. But VS Code is ignoring after the `.` (as the syntax is actually `<tokentype>.<modifier>` in VS Code, no LSP). So if the first name happened to be in the list (or we used something like `string`) it accidentally worked. But `constant.numeric` (or `constant`) doesn't exist as a TokenType, so it gets no highlighting applied. A user cannot just do `@Category{number}` instead, as rascal-lsp has a hardcoded list of token types, and `number` is not in there.

**Solution**

1. replace the list with the enum form the LSP spec.
2. consider adding a mapping of the old text-mate styles to the token types, so that existing languages keep working a reasonable manner.

### PR DESCRIPTION
**Pull request by DavyLandman**
The original implementation [was half working, due to an incorrect interpretation of the standard](#366). This commits fixes that. For backwards compatibility: it still supports the legacy ones from rascal and also the wrong text-mate ones.

The rendering of Rascal is the same, except that string and numeric literals are now both highlighted as string (where before they were not highlighted). This is something we can only fix in the rascal grammar.

Also, side-effect, the list of tokens send to the client is less, as we only try and send those that have an `category` and not all the plain ones.

Fixes #366

TODO:

- [x] extend documentation for DSL
- [ ] (not part of this PR) tune categories in Rascal grammar

### CODE REVIEW

**Comment by DavyLandman** [comment]
TODO:

- [ ] test modifiers in a DSL
- [ ] extend documentation for DSL
- [ ] (not part of this PR) tune categories in Rascal grammar

**Review by urbanfly — COMMENTED** [review]
Neat to see how this works under-the-hood.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java [hunk 1]
```diff
@@ -95,294 +100,316 @@ public SemanticTokensCapabilities capabilities() {
...
+        private static String[][] backwardsCompatibleTokenTypes = new String[][] {
+            /**
+             * The Rascal legacy token types are translated to
+             * textmate token types here
+             */
+            //{ TreeAdapter.NORMAL,               null },
+            { TreeAdapter.TYPE,                 SemanticTokenTypes.Type},
+            { TreeAdapter.IDENTIFIER,           SemanticTokenTypes.Variable},
+            { TreeAdapter.VARIABLE,             SemanticTokenTypes.Variable },
+            { TreeAdapter.CONSTANT,             SemanticTokenTypes.String }, // it's a choice between number and string, both are wrong
```
[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java:138]: no, as the point is that it's a constant. so they are often string literals of numerics. so either on of them would be test. so I opted for strings as they are maybe a bit better.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java [hunk 2]
```diff
@@ -95,294 +100,316 @@ public SemanticTokensCapabilities capabilities() {
...
+            {"keyword.declaration", SemanticTokenTypes.Modifier },
+            {"keyword.operator.arithmetic", SemanticTokenTypes.Operator },
+            {"keyword.operator.assignment", SemanticTokenTypes.Operator },
+            {"keyword.operator.bitwise", SemanticTokenTypes.Operator },
+            {"keyword.operator.logical", SemanticTokenTypes.Operator },
+            {"keyword.operator.word", SemanticTokenTypes.Operator },
+            {"keyword.operator", SemanticTokenTypes.Operator },
+            {"keyword.other", SemanticTokenTypes.Keyword },
+            {"keyword", SemanticTokenTypes.Keyword },
+            // {"markup.bold", null },
```
[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java:255]: These used to be part of the original definition of tokens supported, but there is no clear mapping to the semantic tokens, so if some user used that (which I doubt) they'll loose syntax highlighting. I could collect them in a single place. that just makes the diff bigger (if you use `word-diff` not line diff).

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java [hunk 3]
```diff
@@ -95,294 +100,316 @@ public SemanticTokensCapabilities capabilities() {
...
+            {"constant.numeric.integer.binary", SemanticTokenTypes.Number },
+            {"constant.numeric.integer.decimal", SemanticTokenTypes.Number },
+            {"constant.numeric.integer.hexadecimal", SemanticTokenTypes.Number },
+            {"constant.numeric.integer.octal", SemanticTokenTypes.Number },
+            {"constant.numeric.integer.other", SemanticTokenTypes.Number },
+            {"constant.numeric.integer", SemanticTokenTypes.Number },
+            {"constant.numeric", SemanticTokenTypes.Number },
+            {"constant.other.placeholder", SemanticTokenTypes.Number },
+            {"constant.other", SemanticTokenTypes.Number },
+            {"constant", SemanticTokenTypes.Number },
```
[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java:209]: damn, and I thought I had all the duplicates already, there were a few more. thanks for spotting.

File: rascal-vscode-extension/package.json [hunk 4]
```diff
@@ -138,6 +138,27 @@
         "path": "./syntaxes/parametric.tmGrammar.json"
       }
     ],
+    "semanticTokenScopes": [
```
Comment (urbanfly) [inline rascal-vscode-extension/package.json:141]: ~Can you summarize what this section does?~

---

Actually, I decided to ask GitHub Copilot Chat about this snippet of code with no other context. Is it right?

> In a vs code extension, what does this block mean?

<details><summary>The answer</summary>
<p>
This JSON block is part of a Visual Studio Code extension's configuration, specifically related to semantic highlighting.

Semantic highlighting is a feature that provides more granular syntax coloring based on the semantic context, rather than just the type of the token. This can make it easier to understand the structure and meaning of your code.

Here's a breakdown of the JSON:

- `"semanticTokenScopes"`: This is an array of objects, each defining the semantic token scopes for a specific language. In this case, two languages are defined: `rascalmpl` and `parametric-rascalmpl`.

  - `"language"`: The language identifier.
  - `"scopes"`: An object mapping semantic token types to textmate scopes. In this case, the `ambiguity` token type is mapped to the `invalid` textmate scope.

- `"semanticTokenTypes"`: This is an array of objects, each defining a semantic token type.

  - `"id"`: The identifier of the semantic token type.
  - `"superType"`: The super type of this semantic token type. This is used for inheritance of properties.
  - `"description"`: A human-readable description of the semantic token type.

In this case, the `ambiguity` token type is defined, which is used to highlight ambiguous parts of a parse tree in the Rascal language. The `invalid` scope is typically used to highlight errors or problems in code, so it seems this extension uses it to highlight these ambiguous parts.
</p>
</details>

**Comment by urbanfly** [comment]
Will this PR make syntax highlighting Rascal code itself better/different? (like numbers in Rascal code will now be distinct?)

File: rascal-vscode-extension/package.json (hunk 4 above)
Comment (DavyLandman) [reply rascal-vscode-extension/package.json:141]: It's documented in these two sections: 
- https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#custom-token-types-and-modifiers 
- https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#custom-textmate-scope-mappings

**Comment by DavyLandman** [comment]
> Will this PR make syntax highlighting Rascal code itself better/different? (like numbers in Rascal code will now be distinct?)

Yes, they will now be highlighted (since before they were just the regular text color). But it requires some further tuning to get a difference between the string literals and the numeric literals.

File: rascal-lsp/src/main/rascal/util/LanguageServer.rsc [hunk 5]
```diff
@@ -63,6 +63,14 @@ Language language(PathConfig pcfg, str name, str extension, str mainModule, str
     = language(pcfg, name, {extension}, mainModule, mainFunction);
 
 @synopsis{Function profile for parser contributions to a language server}
+@description{
+    The parser function takes care of parsing the tree once after every change in the IDE.
+    This parse tree is then used for both syntax highlighting and other language server functions.
+}
+@pitfalls {
+    * use ((function::ParseTree::parser)) instead of writing your own function to ensure syntax highlighting is fast
```
[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/rascal/util/LanguageServer.rsc]: I've talked to jurgen, it's not working for an current unknown reason.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java]: I've removed these entries.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java [hunk 6]
```diff
@@ -95,294 +100,316 @@ public SemanticTokensCapabilities capabilities() {
...
+            {"entity.name.trait", SemanticTokenTypes.TypeParameter },
+            {"entity.name.type", SemanticTokenTypes.Type },
+            {"entity.name.union", SemanticTokenTypes.Struct },
+            {"entity.name", SemanticTokenTypes.Type },
+            {"entity.other.attribute-name", SemanticTokenTypes.Decorator },
+            {"entity.other.inherited-class", SemanticTokenTypes.Class },
+            {"entity", SemanticTokenTypes.Type },
+            //{"invalid.deprecated", null },
+            //{"invalid.illegal", null },
+            //{"invalid", null },
```
[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java]: tnx, removed the commented entries.
